src/Features/CSharp/Portable/CodeFixes/UseImplicitTyping/UseExplicitTypingCodeFixProvider.cs

```diff
@@ -0,0 +1,97 @@
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.UseImplicitTyping
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.UseExplicitTyping), Shared]
+    internal class UseExplicitTypingCodeFixProvider : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds =>
+            ImmutableArray.Create(IDEDiagnosticIds.UseExplicitTypingDiagnosticId);
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var document = context.Document;
+            var span = context.Span;
+            var root = await document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+
+            var token = root.FindToken(span.Start);
+            if (!token.Span.IntersectsWith(span))
+            {
+                return;
+            }
+
+            var node = token.GetAncestors<SyntaxNode>().First(n => n.Span.Contains(span));
+            if (node == null ||
+                !(node.IsKind(SyntaxKind.PredefinedType) ||
+                  node.IsKind(SyntaxKind.ArrayType) ||
+                  node.IsKind(SyntaxKind.IdentifierName) ||
+                  node.IsKind(SyntaxKind.GenericName) ||
+                  node.IsKind(SyntaxKind.AliasQualifiedName)))
+            {
+                return;
+            }
+
+            var codeAction = new MyCodeAction(
+                CSharpFeaturesResources.UseExplicitTyping,
+                c => HandleDeclaration(document, root, node, context.CancellationToken));
+
+            context.RegisterCodeFix(codeAction, context.Diagnostics.First());
+        }
+
+        private async Task<Document> HandleDeclaration(Document document, SyntaxNode root, SyntaxNode node, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var declarationContext = node.Parent;
+
+            TypeSyntax typeSyntax;
+            if (declarationContext is VariableDeclarationSyntax)
+            {
+                typeSyntax = ((VariableDeclarationSyntax)declarationContext).Type;
+            }
+            else if (declarationContext is ForEachStatementSyntax)
+            {
+                typeSyntax = ((ForEachStatementSyntax)declarationContext).Type;
+            }
+            else
+            {
+                Debug.Assert(false, $"unhandled kind {declarationContext.Kind().ToString()}");
+                return null;
+            }
+
+            var typeSymbol = semanticModel.GetTypeInfo(typeSyntax).ConvertedType;
+
+            var typeName = SyntaxFactory.ParseTypeName(typeSymbol.ToDisplayString())
```

Comment:
@DustinCampbell : Looks like `ParseTypeName` consumes unexpected tokens in the given string and sets `diagnostics`. I'm thinking of asserting that the parsed name does not contain diagnostics.  there is also a `TypeSyntaxExtensions.IsPotentialTypeName` which does some syntactic checks and some semantic checks as well. However, it feels a bit too much to me, since `ParseTypeName` seems to attach diagnostics to the node and it seems sufficient. Does that sound good?